### PR TITLE
container: Fix container semantics

### LINKED ISSUE #588: the container returned by createContainer must be the same one contained in pod.containers
in order to avoid semantic errors modifying the container state of two containers with the same container ID, the container returned by `createContainer` must be the same one contained in `pod.containers`

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request ensure the container pointers held by the `pod` structure are not duplicated. This is needed to avoid a container pointer modification not to be taken into account because the wrong pointer is used.

Fixes #588

### CODE REVIEW

**Comment by sboeuf** [comment]
cc @devimc @sameo

**Comment by jodh-intel** [comment]
Hi @sboeuf - please can you add some unit tests for this change.

**Comment by devimc** [comment]
lgtm

**Comment by sboeuf** [comment]
@jodh-intel I have added unit tests in order to cover `fetchContainer()` function since this is the one that matters here. It makes sure we retrieve the container pointer coming from the list of containers held by the pod structure.

File: container.go [hunk 1]
```diff
@@ -244,18 +244,13 @@ func fetchContainer(pod *Pod, containerID string) (*Container, error) {
...
-	if err != nil {
-		return nil, fmt.Errorf("failed to create container with config %v in pod %v: %v", config, pod.id, err)
+	for _, c := range pod.containers {
+		if containerID == c.id {
+			return c, nil
+		}
 	}
 
-	return container, nil
+	return nil, fmt.Errorf("Could not find the container in the pod's containers list")
```
Comment (jodh-intel) [inline container.go]: Could you add `containerID` to the returned error (and maybe `pod.ID()`)?

Comment (sboeuf) [reply container.go]: Oh yeah sorry I forgot that ! Done !

**Comment by jodh-intel** [comment]
Thanks @sboeuf.

lgtm

**Comment by egernst** [comment]
Nice.

LGTM

File: container.go [hunk 2]
```diff
@@ -244,18 +244,14 @@ func fetchContainer(pod *Pod, containerID string) (*Container, error) {
...
-		return nil, err
-	}
-
-	container, err := createContainer(pod, config)
-	if err != nil {
-		return nil, fmt.Errorf("failed to create container with config %v in pod %v: %v", config, pod.id, err)
+	for _, c := range pod.containers {
+		if containerID == c.id {
+			return c, nil
+		}
```
Comment (sameo) [inline container.go]: A couple of things on that one:

* This needs a comment.
* Having to do so means we're fetching a container from storage while having a fully built pod with all its container already fetched and added. When is that happening ?

**Comment by sboeuf** [comment]
@sameo comments addressed, I have replaced the function `fetchContainer()` with `findContainer()`, which is more appropriate about the naming.

**Comment by sameo** [comment]
LGTM

**Pull request merged**